### PR TITLE
feat: getPersonas and callAsPersona

### PR DESCRIPTION
**Pull request by DanThomp507**

### CODE REVIEW

**Comment by DanThomp507** [comment]
[[SDK] personas work](https://trello.com/c/WdWnfbOp/1095-sdk-personas-work)

File: src/pod-utils.ts [hunk 1]
```diff
@@ -0,0 +1,20 @@
...
+// Functions are put here so that they can be mocked for testing purposes.
+
+// eslint-disable-next-line import/prefer-default-export
+export async function getPersonas(
+  pod: Pod,
+  address: string,
+): Promise<Array<{ type: string; address: string }>> {
+  const personas = [];
+  if (pod.isAdmin(address)) personas.push({ type: 'admin', address });
+  if (await pod.isAdminPodMember(address)) personas.push({ type: 'adminPodMember', address });
```
Comment (jtlacci) [inline src/pod-utils.ts:12]: this this should be `address: adminPod.address`

Comment (jtlacci) [reply src/pod-utils.ts:12]: i think lets validate that

**Review by jtlacci — APPROVED** [review]

File: test/fixtures/index.ts [hunk 2]
```diff
@@ -59,6 +59,8 @@ export const artNautPod = {
   ],
 };
 
+export const gmPod = {};
+
```
Comment (kevinweaver) [inline test/fixtures/index.ts:63]: Looks like you didn't end up using this

File: test/pod.persona.test.ts [hunk 3]
```diff
@@ -0,0 +1,229 @@
...
+        toNumber: jest.fn().mockImplementation(() => 10),
+      }),
+    },
+    podId: expectedPod.id,
+    Name: { name: expectedPod.name },
+  });
+}
+
+const provider = new ethers.providers.InfuraProvider('goerli', {
+  infura: '69ecf3b10bc24c6a972972666fe950c8',
```
Comment (kevinweaver) [inline test/pod.persona.test.ts:41]: ⚠️ Doesn't need to block this PR, but do you mind moving this to an env var for this repo as well? I see 10 instances of it: https://github.com/search?q=org%3Aorcaprotocol+69ecf3b10bc24c6a972972666fe950c8&type=code

Is this a development key or do we use it in prod as well?

@heyitschun Do you mind adding a low priority backlog ticket to eventually rotate this infura key? It's not the end of the world if it leaks, but it's good hygiene and practice for us to separate prod keys regardless.

File: test/pod.persona.test.ts [hunk 4]
```diff
@@ -0,0 +1,229 @@
...
+      type: 'subPodMember',
+      address: '0x203D96211d5f0e0De9788302505dd37D26a28005',
+    },
+  ]);
+  const personas = await pod.getPersonas('0x08A9C4898e045A4E176C2cbf4Ca86aEf9f6EA478');
+
+  await expect(pod.callAsPersona(pod.mintMember, [userAddress2], personas[0])).rejects.toThrow(
+    'Sub pod was not a member of this pod',
+  );
+});
```
Comment (kevinweaver) [inline test/pod.persona.test.ts:229]: 🤔 Do you mind breaking these tests up by each persona type?
Ex: move line 51-104 into a describe block, like 
```
describe('admin member', () => {
...
})
```

File: test/pod.persona.test.ts [hunk 5]
```diff
@@ -0,0 +1,229 @@
...
+  };
+
+  const mockMint = jest.fn();
+  jest.spyOn(utils, 'getMetropolisContract').mockReturnValueOnce({
+    mint: mockMint,
+  });
+
+  await expect(
+    pod.callAsPersona(pod.mintMember, [expectedPod.admin], personas[0], mockSigner),
+  ).rejects.toThrow('Signer was not admin');
```
Comment (kevinweaver) [inline test/pod.persona.test.ts:103]: ❓ Where is this error actually thrown in the code? I don't see it in the PR, so I assume it's not part of the code changes?

File: test/pod.persona.test.ts [hunk 6]
```diff
@@ -0,0 +1,229 @@
...
+// Some of these tests run long.
+jest.setTimeout(7500);
+
+const expectedPod = {
+  admin: '0xf0C7d25c942264D6F21871c05d3dB3b98344b499',
+  id: 6,
+  name: '1-member-pods.pod.eth',
+  safe: '0xC99cE75a83B068DABF48c48f91C62a30fe0afBd2',
+  member: '0x61De0bbb6C8215Af3f821FE4884A28bc737f98D3', // random member
+};
```
Comment (kevinweaver) [inline test/pod.persona.test.ts:18]: 🤔 Do you mind setting variables for all of these addresses? And maybe for different personal types?  Ex:
`const adminAddress = '0xf0C7d25c942264D6F21871c05d3dB3b98344b499'`

I think it'll make the tests easier to follow.

File: test/pod.persona.test.ts [hunk 7]
```diff
@@ -0,0 +1,229 @@
...
+  const personas = await pod.getPersonas('0x85760ef61c0ccB7BCC4C7A0116d80D59D92e736d');
+
+  await pod.callAsPersona(
+    pod.mintMember,
+    [userAddress2],
+    personas[0],
+    '0x85760ef61c0ccB7BCC4C7A0116d80D59D92e736d',
+  );
+  expect(create).toHaveBeenCalled();
+});
```
Comment (kevinweaver) [inline test/pod.persona.test.ts:155]: Example of a test that would be easier to follow with variable names that explicitly show what persona they are.

**Review by kevinweaver — COMMENTED** [review]
Mostly questions. I vaguely recall this work being discussed, but I don't have much context so it was a little hard to follow. The PR description is empty and the ticket just says "follow up work from" another ticket, which also has no description. 😅 

I know we've talked about increasing ticket description detail, but @heyitschun do you mind adding a ticket to add a PR template to the SDK as well? That will at least prompt us to write in something.

File: src/Pod.ts [hunk 8]
```diff
@@ -418,6 +427,99 @@ export default class Pod {
...
+   */
+  callAsPersona = async (
+    method: any,
+    args: Array<any>,
+    persona: { type: string; address: string },
+    sender?: ethers.Signer | string,
+  ) => {
+    switch (persona.type) {
+      case 'admin':
+        if (!sender) throw new Error(`Expected sender to be signer, but received ${sender}`);
```
[the remark this answers is not in the document]
Comment (jtlacci) [reply src/Pod.ts:459]: this is the only case where we are actually making a transaction, all other cases we are making proposals

File: src/pod-utils.ts (hunk 1 above)
Comment (DanThomp507) [reply src/pod-utils.ts:12]: yeah you're spot on, have updated this now and will push up in a bit!

File: test/pod.persona.test.ts (hunk 6 above)
Comment (DanThomp507) [reply test/pod.persona.test.ts:19]: Of course! Don't mind at all. Will update now.

File: test/pod.persona.test.ts (hunk 5 above)
Comment (DanThomp507) [reply test/pod.persona.test.ts:104]: This is an error within the mint member function so that's already been set up!

**Review by kevinweaver — APPROVED** [review]
Thanks for addressing my comments!
